### ISSUE TITLE
JSON formatter outputs invalid JSON object due to "\n" in source property

### ISSUE THREAD

**Issue by rchl** [eslint/eslint#7523]
**Tell us about your environment**

* **ESLint Version: v3.8.1**
* **Node Version: v5.10.1**
* **npm Version: 3.10.9**

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
`eslint --no-eslintrc --rule 'comma-dangle:[error,always-multiline]' --format json`

**What did you do? Please include the actual source code causing the issue.**
```js
var foo = {
  bar: "baz",
  qux: "quux"
};
```

**What did you expect to happen?**
The returned object t be a valid JSON object.

**What actually happened? Please include the actual, raw output from ESLint.**
The `source` property of the JSON object contains literal `\n` characters which are not valid in a JSON property value. It must be double-escaped (`\\n`) to be valid.
Invalid output:
```json
[{"filePath":"D:\\workspace\\work\\test.js","messages":[{"ruleId":"comma-dangle","severity":2,"message":"Missing trailing comma.","line":3,"column":14,"nodeType":"Property","source":"  qux: \"quux\"","fix":{"range":[39,39],"text":","}}],"errorCount":1,"warningCount":0,"source":"var foo = {\n  bar: \"baz\",\n  qux: \"quux\"\n};\n"}]
```

**Comment by platinumazure** [comment]
I'm a little confused by this report. The JSON you are showing seems to be valid (you are showing unquoted JSON, so I would expect `\n` to be interpreted as a backslash-n rather than an actual newline.

@vitorbal also took a look and fed your JSON through a validator, and it says it's valid.

How are you using the JSON? I'm wondering if whatever tool you are feeding the JSON incorrectly into a tool, or maybe the tool has a bug deserializing the JSON?

**Comment by rchl** [comment]
I'm feeding it to python's JSON parser directly:
```
% python
Python 2.7.10 (default, Jun  1 2015, 18:05:38)
[GCC 4.9.2] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> json.loads('[{"filePath":"D:\\workspace\\work\\test.js","messages":[{"ruleId":"comma-dangle","severity":2,"message":"Missing trailing comma.","line":3,"column":14,"nodeType":"Property","source":"  qux: \"quux\"","fix":{"range":[39,39],"text":","}}],"errorCount":1,"warningCount":0,"source":"var foo = {\n  bar: \"baz\",\n  qux: \"quux\"\n};\n"}]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
    obj, end = self.scan_once(s, idx)
ValueError: Invalid \escape: line 1 column 17 (char 16)
```

But I think you are right, I was confused due to python's string parsing. The output itself seems fine.

Although I'm not quite sure how to handle that in python properly. Output is obtained through a subprocess call so I can't just define it as raw string. Using string.encode('string-escape') doesn't seem to work correctly for that input either.
That is more of a support question rather than the bug so sorry and you can close this I think.

**Comment by platinumazure** [comment]
At this point, I'd say we don't have a bug.

Re: feeding the string in, try using a raw string (`r'here is a raw string, just prefix the literal with an r'`) or reading from a file/pipe to avoid the parse issue. (Of course, this isn't a Python project so I don't think we can provide much more support than that here.)

Good luck!

**Comment by not-an-aardvark** [comment]
As an alternative, you could try double-escaping all the backslashes so that they don't get interpreted as python special characters when you paste them into a string:

```
eslint --format json foo.js | sed 's/\\/\\\\/g'
```

**Comment by nzakas** [comment]
Closing, as this doesn't appear to be an issue with ESLint.
